src/reducers/console.js

```diff
@@ -35,20 +33,62 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
+      return payload.trim(' ') === '' ? state : state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({expression: payload}),
       );
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
-        new ConsoleEntry({
+      return state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
+      );
+
+      const newHistoryEntryIndex = payload.direction === 'UP' ?
+        state.historyEntryIndex + 1 :
+        state.historyEntryIndex - 1;
```

Comment:
It would be pretty slick to define an [enum](https://github.com/popcodeorg/popcode/blob/master/src/enums/index.js) with explicit values for `UP` and `DOWN` (which the [`enum` library](https://www.npmjs.com/package/enum) lets you do by initializing the enum with an object) so that we can just call `payload.direction.value` here rather than having to do a conditional.

Reply:
I do like enums but I think the real refactoring win for this code would be to split up and down navigation into different actions. That way, there's no need for a conditional to figure out if the index should be incrementing or decrementing. It would also mean we don't have to use constant strings like 'UP' nor enums.

If you agree that would be beneficial, such an approach would necessitate a shared function that does the history lookup (splitting out this part is itself a good step IMO, since the `case 'NAVIGATE_CONSOLE_HISTORY':` is already 38 lines). Could I put such a function in the top level of the `src/reducers/console.js` module (not being exported)?